[RUM-2158] Document more flexible proxy URL (#20720)
DataDog/documentation: content/en/real_user_monitoring/guide/proxy-rum-data.md

## Overview

The RUM Browser SDK can be configured to send requests through a proxy. Requests made still need to be forwarded to Datadog.
When you set the `proxy` [initialization parameter][1], all RUM data is sent to the specified URL using the POST method (for example, `https://www.proxy.com/foo`).

## Prerequisites

> A request to datadog (ex: `https://browser-intake-datadoghq.eu/api/v2/rum?ddsource=browser&...`) is composed of three parts:

Reviewer: ```suggestion
A Datadog request URL (ex: `https://browser-intake-datadoghq.eu/api/v2/rum?ddsource=browser&...`) is composed of three parts:
```

- The Datadog intake origin corresponding to your `site` [initialization parameter][1] (ex: `https://browser-intake-datadoghq.eu`)
- The path containing the API version and the product (ex: `/api/v2/rum`)
- The parameters (ex: `ddsource=browser&...`)

> The Datadog intake origin corresponding to your site parameter should be defined in your proxy implementation. The path and parameters should be attached to the requests to the proxy.

Reviewer 2: ```suggestion
The Datadog intake origin corresponding to your site parameter should be defined in your proxy implementation. The path and parameters should be appended to the requests sent to the proxy.
```

Datadog intake origins for each site are listed below:

| Site    | Site Parameter            | Datadog intake origin                      |
|---------|---------------------------|--------------------------------------------|
| US1     | `datadoghq.com` (default) | `https://browser-intake-datadoghq.com`     |
| US3     | `us3.datadoghq.com`       | `https://browser-intake-us3-datadoghq.com` |
| US5     | `us5.datadoghq.com`       | `https://browser-intake-us5-datadoghq.com` |
| EU1     | `datadoghq.eu`            | `https://browser-intake-datadoghq.eu`      |
| US1-FED | `ddog-gov.com`            | `https://browser-intake-ddog-gov.com`      |
| AP1     | `ap1.datadoghq.com`       | `https://browser-intake-ap1-datadoghq.com` |

> To successfully proxy request to Datadog:

Reviewer: ```suggestion
To successfully proxy a request to Datadog:
```

1. Build the final Datadog intake URL by using:
   - the Datadog intake origin corresponding to your site.

>    - the path and parameters from the request to the proxy.

Reviewer 2: ```suggestion
   - the path and parameters from the request sent to the proxy.
```

>    - the Datadog intake origin corresponding to your site.
>    - the path and parameters from the request to the proxy.

Reviewer: Those lines are missing one space to be properly indented in the output HTML

```suggestion
    - the Datadog intake origin corresponding to your site.
    - the path and parameters from the request to the proxy.
```

2. Add a `X-Forwarded-For` header containing the request client IP address for accurate geoIP.
3. Forward the request to the Datadog intake URL using the POST method.
4. Leave the request body unchanged.

> The request body can contain binary data and should not be converted to a string. Make sure your proxy implementation forwards the raw body without conversion.

Reviewer 2: <!-- Suggestions are specific requests to improve the subject under review. It is assumed that we all want to do what's best, so these comments are never dismissed as “mere suggestions”, but are taken seriously.  -->
**💬 suggestion:**  It could be nice to highlight this sentence as it's something error prone in proxy config (and even more now since we will compress RUM data)

});
```

{{% /tab %}}
{{% tab "CDN async" %}}
```javascript
window.DD_RUM.onReady(function() {
    window.DD_RUM.init({
        clientToken: '<CLIENT_TOKEN>',
        applicationId: '<APPLICATION_ID>',
        proxy: (options) => `https://www.proxy.com/foo${options.path}/bar?${options.parameters}`,
    })
})
```
{{% /tab %}}
{{% tab "CDN sync" %}}

```javascript
window.DD_RUM &&
    window.DD_RUM.init({
        clientToken: '<CLIENT_TOKEN>',
        applicationId: '<APPLICATION_ID>',
        proxy: (options) => `https://www.proxy.com/foo${options.path}/bar?${options.parameters}`
    });
```

{{% /tab %}}
{{< /tabs >}}

For example, with a `site` set to `datadoghq.eu` and the `proxy` configuration from the example, the RUM Browser SDK sends requests to an URL that looks this: `https://www.proxy.com/foo/api/v2/rum/bar?ddsource=browser`. The proxy will need to forward the request to the URL `https://browser-intake-datadoghq.eu/api/v2/rum?ddsource=browser`.

**Note:** Some privacy blockers already target the intake [url patterns][2], so you may want to take that into account when building your proxy URL.

> **Note:** The `proxy` function is called for each request, so it should avoid any heavy computation.

Reviewer 2: I don't know why but the notes are not displayed as expected:
![image](https://github.com/DataDog/documentation/assets/4342515/8851ae98-4bd0-4293-9fff-d40c6f997d95)

Reviewer: Markdown expect a blank line between paragraphs
